### PR TITLE
Fix arange issue for int8/int16/int32 dtype

### LINKED ISSUE #974: [torchlib] Fix arange for int32
Output shape mismatch in edge cases.

### PR DESCRIPTION
**Pull request by fatcat-z**
ONNX op 'Range' behaves differently with Torch op 'arange' when the expected output dtype is INT32.

Totally 2 differences are found in two conditions below:

1. 'start' is negative, or 
2. 'step' is negative. 

This PR will make necessary adjustments to make these 2 ops behave same.

Fix #974 .

### CODE REVIEW

**Comment by justinchuby** [comment]
Thanks! We can work together to simplify the logic on Wednesday. For now - should we make this compatible with int16 and int8 as well, since they should be the same as int32?

**Comment by fatcat-z** [comment]
> Thanks! We can work together to simplify the logic on Wednesday. For now - should we make this compatible with int16 and int8 as well, since they should be the same as int32?

Has updated the code to support them as well. It passed in local testing.

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -484,6 +484,28 @@ def aten_arange_start(
...
+    start: TRealUnlessFloat16OrInt8,
+    end: TRealUnlessFloat16OrInt8,
+    step: TRealUnlessFloat16OrInt8,
+    dtype: int,
+) -> Tuple[FLOAT, FLOAT, FLOAT]:
+    zero = op.Cast(0.0, to=FLOAT.dtype)
+    start = op.Cast(start, to=FLOAT.dtype)
+
+    if start < zero:
+        start = op.Cast(op.Cast(start, to=dtype), to=FLOAT.dtype)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:498]: Is this effectively taking Floor?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 2]
```diff
@@ -484,6 +484,28 @@ def aten_arange_start(
...
 
+@torch_op("aten::arange.start_step", private=True)
+def _aten_arange_start_step_onnx(
+    start: TRealUnlessFloat16OrInt8,
+    end: TRealUnlessFloat16OrInt8,
+    step: TRealUnlessFloat16OrInt8,
+    dtype: int,
+) -> Tuple[FLOAT, FLOAT, FLOAT]:
+    zero = op.Cast(0.0, to=FLOAT.dtype)
+    start = op.Cast(start, to=FLOAT.dtype)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:495]:
```suggestion
    start = op.Cast(start, to=FLOAT.dtype)
    step = op.Cast(step, to=FLOAT.dtype)
    end = op.Cast(end, to=FLOAT.dtype)
```

We can move all casts to the top for clarity

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 3]
```diff
@@ -484,6 +484,28 @@ def aten_arange_start(
     return result
 
 
+@torch_op("aten::arange.start_step", private=True)
+def _aten_arange_start_step_onnx(
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:488]:
```suggestion
def _adjust_endpoints_for_arange_int_dtype(
```

or something more descriptive?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 4]
```diff
@@ -499,18 +521,26 @@ def aten_arange_start_step(
     if dtype == -1:
         result = op.Range(start, end, step)
     elif _range_supported(dtype):
-        end = op.Cast(end, to=dtype)
-        start = op.Cast(start, to=dtype)
-        step = op.Cast(step, to=dtype)
-        result = op.Range(start, end, step)
+        if dtype in [INT16.dtype, INT32.dtype]:
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:524]:
```suggestion
        if dtype in (INT8.dtype INT16.dtype, INT32.dtype):
```

I would also use an immutable type (tuple) for immutable uses

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 5]
```diff
@@ -499,18 +521,26 @@ def aten_arange_start_step(
...
+            result = op.Range(start, end, step)
     else:
         # Cast input to float if dtype is not supported by Range,
         # because the input dtype may be e.g. bfloat16 / int8 etc.
         # which Range does not support. The output type is ensured because the output
         # is casted to the specified dtype.
-        end = op.Cast(end, to=FLOAT.dtype)
-        start = op.Cast(start, to=FLOAT.dtype)
-        step = op.Cast(step, to=FLOAT.dtype)
+        if dtype == INT8.dtype:
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:537]: Looks like we can merge this with the other int types? Only INT64 is different

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:498]: Sorry my bad - I should say Ceil

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 6]
```diff
@@ -484,6 +484,28 @@ def aten_arange_start(
     return result
 
 
+@torch_op("aten::arange.start_step", private=True)
+def _aten_arange_start_step_onnx(
+    start: TRealUnlessFloat16OrInt8,
+    end: TRealUnlessFloat16OrInt8,
+    step: TRealUnlessFloat16OrInt8,
+    dtype: int,
+) -> Tuple[FLOAT, FLOAT, FLOAT]:
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:493]: Note: comment that links to the pytorch ref implementation that explains this logic

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 7]
```diff
@@ -425,6 +425,15 @@ def _range_supported(dtype: int) -> bool:
     }
 
 
+def _integral_to_be_adjusted(dtype: int) -> bool:
```
Comment (titaiwangms) [inline onnxscript/function_libs/torch_lib/ops/core.py:428]: Should we have another file for utils?

**Review by titaiwangms — APPROVED** [review]

**Review by justinchuby — APPROVED** [review]
Thanks for the great work!

**Pull request merged**